Added support for Node-Restify

theoutlander opened the pull request:
There were recent changes to their API in their documentation that I've accounted for the most part. If I've missed anything, I'll add it as I discover it. For now, I think this is a good starting point and we can continue to add missing stubs. Let me know if there are any fixes to this code that I should make. Thanks!

theoutlander committed:
Added support for Node-Restify

Applied changes as suggested in the comments in https://github.com/nikhilk/scriptsharp/pull/353

Applied changes as suggested in the comments in https://github.com/nikhilk/scriptsharp/pull/353

Added support for Node-Restify

Additional fixes to initial Node-restify library port

Removing files for initial node-restify commit

Removing files for initial node-restify commit

Created node-restify stubs for s#

theoutlander committed:
Added support for Node-Restify

nikhilk commented:
I am going to comment inline as I look through the code next. Hopefully you'll be able to git squash at the end to cleanup the actual pull request.

nikhilk commented on src/Libraries/Node/Node.Core/Network/HttpServerRequest.cs:
```diff
@@ -11,9 +11,9 @@ namespace NodeApi.Network {
     [ScriptImport]
     [ScriptIgnoreNamespace]
-    public sealed class HttpServerRequest : ReadableStream, IEventEmitter {
+    public class HttpServerRequest : ReadableStream, IEventEmitter {
```
I didn't notice something in the restify docs to document the restify request deriving from the core http equivalents.

Independent of that though, I feel it might be better to simply duplicate, since:
- Not necessary everything on HttpServerRequest should be on the RestifyRequest.
- This opens up derivation possibilities in script# code which would neither be right, nor supported.

Same applies to HttpServerResponse.

Might be best duplicating the relevant properties.

theoutlander replied:
In [Restify's Request API section](http://mcavage.github.io/node-restify/#Request-API), it says that it wraps all of the node ServerRequest API's, events and properties, plus the Restify specific entities. 

I don't think we should be replicating functionality as we'll have to keep them in sync every time there are changes in Node's Request object. Same goes for Response.

I can duplicate for now, although I have not been successful at using git squash and have previous changes discarded after sending a pull request. Earlier I had to drop my repo and re-fork  to get rid of some of the things. If you have any insights into how to do it, that'll be helpful.

theoutlander replied:
I don't think there's a way to undo changes/history for a specific file from an existing push.
I'm assuming that by using git commit --squash, you want me to bundle all the changes in a single request.

nikhilk commented on src/Libraries/Node/Node.Restify/BunyanLogger.cs:
```diff
@@ -0,0 +1,18 @@
+    
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    public sealed class BunyanLogger {
```
Mostly questions since I don't know about this.

Is this a restify module? Should it be implementing some interface so other potential restify modules can be modeled in script# code?

Shouldn't it be bringing its own dependency, i.e, if it requires its own "require" call be emitted into the script?

Should this be in the Restify namespace?

Does everyone using Restify use Bunyan logger? Or can they plug it in? If its a choice, and not a mainstream one, it might suggest this shouldn't be in the Restify assembly. The reason I ask is it seems odd that I'd call this API directly from app code - this seems like background infrastructure/middleware that gets registered with restify.

theoutlander replied:
This is a separate module. Just noticed that it's optional, so I can remove it for now and leave the property as an object. I think if we supported Bunyan logger in the future, we could reference it.

nikhilk commented on src/Libraries/Node/Node.Restify/Node.Restify.vsdoc:
```diff
@@ -0,0 +1,7 @@
```
Not sure what this file is - some specific to your environment? Can you remove?

theoutlander replied:
Oh that's right. Not sure why it keeps creating it. I'll remove it.

nikhilk commented on src/Libraries/Node/Node.Restify/Restify.cs:
```diff
@@ -0,0 +1,80 @@
+using System;
+using System.Runtime.CompilerServices;
+
+namespace NodeApi.Restify {
```
I am assuming we want to keep the class named as Restify, so that would mean we want a different/non-conflicting namespace name.

It was straightforward in expressjs since the module is called ExpressJS, leaving the Express name free.

I am not thrilled about these alternatives, but maybe they trigger some other better alternative. RestifyServer, RestifyService, or worst case, a generic RestifyModule (if we pick the latter, it raises questions on whether other modules should follow suit for consistency). As I write this, RestAPI also comes to mind - I tried to do this with the core nodejs modules - put them in functionally named subnamespaces.

theoutlander replied:
I'll change it to RestifyService since there are Restify client api's (and RestifyServer would be incorrect and confusing in that case). 

RestAPI is not a bad idea either....but I'm not sure about it yet. We can rename later if needed.

nikhilk commented on src/Libraries/Node/Node.Restify/Restify.cs:
```diff
@@ -0,0 +1,80 @@
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    [ScriptName("restify")]
+    public sealed class Restify {
```
This should just be a static class. It has only static methods.

nikhilk replied:
Now that I see the methods in here - maybe this class should be renamed rather than the namespace. This seems equivalent to ExpressApplication in spirit, so maybe this should be called RestifyApplication?

theoutlander replied:
I had thought about doing this but forgot once I was looking into syncing with latest API changes.
I'm going to revert the namespace change and rename Restify to RestifyApplication (to be consistent with ExpressApplication).

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyChain.cs:
```diff
@@ -0,0 +1,26 @@
+        public void Error(Exception error) {
+        }
+    }
+}
```
Just noticed this - could you ensure files end with newline? Its a nit, but it gets rid of this warning that github and some other tools show for files that don't.

theoutlander replied:
Done.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyClient.cs:
```diff
@@ -0,0 +1,14 @@
+
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    public class RestifyClient {
```
abstract?
internal ctor?

theoutlander replied:
Good catch. Meant to use it as a base class, but RestifyStringClient took its place so I'm removing this class.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyError.cs:
```diff
@@ -0,0 +1,14 @@
+    
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    public sealed class RestifyError {
```
internal ctor?

theoutlander replied:
Good catch again. Added internal ctor.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyHttpClient.cs:
```diff
@@ -0,0 +1,17 @@
+
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    public sealed class RestifyHttpClient : RestifyStringClient {
```
I am assuming these classes actually have specific functionality/APIs that hasn't been mapped out yet and eventually will be, hence makes sense to have these classes.

theoutlander replied:
That's correct. I haven't had the need to use them yet so I'll implement their functionality as we get to using it (which helps me validate that the implementation works).

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyJsonClientOptions.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// node-dtrace-provider handle
+        /// </summary>
+        public object Dtrace;
```
This should be cased as DTrace.

theoutlander replied:
Changed to:

[ScriptName("dtrace")]
public object DTrace;

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyRequest.cs:
```diff
@@ -0,0 +1,148 @@
+        /// </summary>
+        [ScriptField]
+        [ScriptName("time")]
+        public int TimeInMs {
```
Might be better fully named TimeInMilliseconds, and adding ScriptName to get to actual script name.

theoutlander replied:
Great point. Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyRequest.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        [ScriptName("accepts")]
+        public string AcceptsContentType(string type) {
```
Does this return bool?

theoutlander replied:
Yes. Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyRequest.cs:
```diff
@@ -0,0 +1,148 @@
+            return String.Empty;
+        }
+
+        public BunyanLogger GetLogger(string name) {
```
Confused - there is a property as well as a logger.

So I noticed this logger is more deeply entrenched (to answer my question from earlier).

Assuming this comes as part of the same module (or its dependency), i.e. doesn't require its own "require" ... shouldn't this just be named RestifyLogger? I imagine you don't create an instance or refer to any static methods so the type name being "bunyan" doesn't help that much, but does raise questions (like what I had).

theoutlander replied:
As for the Log field, each restify request instance has a new bunyan instance log on it where the request id is automatically injected in, so you can easily correlate your high-throughput logs together. The only way to access it is via Log.

As for getLogger(name), it is a shorthand to grab a new bunyan instance that is a child component of the one restify has. BunyanLogger's allow logger names similar to log4j.

I think calling it RestifyLogger is good enough for now. I'll rename the BunyanLogger to reflect that and prevent the user from creating an instance. The calls where a BunyanLogger instance can be passed will be left as object types for now.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyRequest.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        [ScriptName("is")]
+        public string IsContentType(string type) {
```
bool?

theoutlander replied:
Yes, noticed it as I was changing Accepts. Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyResponse.cs:
```diff
@@ -0,0 +1,124 @@
+        /// response headers
+        /// </summary>
+        [ScriptField]
+        public object Headers {
```
Is the type actual `Dictionary<string, string>`?

theoutlander replied:
The type is actually Dictionary<string, object> as values can a primitive or complex type.

nikhilk replied:
Then I'd suggest `Dictionary<string, object>`

theoutlander replied:
Done.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyResponse.cs:
```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// <param name="code">Status code</param>
+        /// <param name="body">content</param>
+        public void Json(HttpStatusCode code, Dictionary body) {
```
This should take in object instead - allowing for dictionaries, generic dictionaries, and record objects. Also applies to Send below.

theoutlander replied:
Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyResponse.cs:
```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// <param name="type">type defaults to _public_</param>
+        /// <param name="options">options currently only takes maxAge.</param>
+        public void SetCache(string type, Dictionary options) {
```
Should this be `Dictionary<string, object>`

theoutlander replied:
Since options only takes maxAge, Dictionary < string, object > is fine. Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServer.cs:
```diff
@@ -0,0 +1,283 @@
+        }
+
+        [ScriptEvent("on", "removeListener")]
+        [ScriptName("MethodNotAllowed")]
```
No need for ScriptName? Same for some other events below...

theoutlander replied:
I've added ScriptName based off the casing specified in their documentation. Leaving it intact for now.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServerOptions.cs:
```diff
@@ -0,0 +1,74 @@
+        ///         }
+        ///     });
+        /// </example>
+        public string ResponseTimeFormatter;
```
From the example above, this should be a func of the right signature.

theoutlander replied:
Good catch. Wonder where I picked up string from. Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyStringClient.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <param name="path"></param>
+        /// <param name="content"></param>
+        /// <param name="callback"></param>
+        public void Post(string path, Dictionary content, RestifyCallback callback) {
```
Dictionary should be object, here and in other similar methods.

theoutlander replied:
Changed all Dictionaries to object.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyThrottleOptions.cs:
```diff
@@ -0,0 +1,83 @@
+
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    public sealed class RestifyThrottleOptions {
```
Is this a vanilla script object at runtime? If so, this likely needs a [ScriptName("Object")] attribute. And to follow the pattern, it implies there should be a ctor that takes a params object[] nameValuePairs arg.

theoutlander replied:
Silly mistake. Changed.

theoutlander commented:
Thanks for all the detailed comments. I'll take a look at these and fix them when I get a chance.

theoutlander committed:
Applied changes as suggested in the comments in https://github.com/nikhilk/scriptsharp/pull/353

theoutlander committed:
# This is a combination of 2 commits.
# The first commit's message is:

Applied changes as suggested in the comments in https://github.com/nikhilk/scriptsharp/pull/353

# This is the 2nd commit message:

Added support for Node-Restify

theoutlander commented:
I've squashed my changes. Hopefully I've done it right. 
I don't know if I should have removed the NodeJS HttpRequest/Response files in the commit... the changes show up as reverted in my check-in.

nikhilk commented on ScriptSharp.sln:
```diff
@@ -0,0 +1,20 @@
+﻿
```
What is this file about?

theoutlander replied:
Weird, no idea where that showed up from. Removing.

nikhilk commented on src/Libraries/Node/Node.Restify/HttpServerRequest.cs:
```diff
@@ -0,0 +1,62 @@
+
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    public class HttpServerRequest : ReadableStream, IEventEmitter {
```
I wasn't suggesting having HttpServerRequest/Response in the Restify namespace, but rather to merge the members on this class into RestifyServerRequest/Response.

theoutlander replied:
Ok, merged those.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyJsonClientOptions.cs:
```diff
@@ -0,0 +1,71 @@
+
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    public sealed class RestifyJsonClientOptions {
```
Should this class have a [ScriptName("Object")] applied to it? Also should it have a parameterless ctor and another ctor accepting a params object[] nameValuePairs parameter?

theoutlander replied:
Yes it needs the script name. Haven't used this class so totally missed that part. 
I didn't realize that Objects can take nameValuePairs and instantiate the members. I had wished there were a way to do it actually, so this is really good to know! 

Changed. Also, changing other options.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyLogger.cs:
```diff
@@ -0,0 +1,18 @@
+    
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    public sealed class RestifyLogger {
```
What are the rules for construction here? Should this be creatable and have a public ctor? Or should it be internal? If the former, what  is the ScriptName that should be specified on this?

theoutlander replied:
That's right, it should actually be private since its instance is returned through other properties and there's no way to create it **for now**. ScriptName should be object - seems like I'm missing a lot of those!

Changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServerGetOptions.cs:
```diff
@@ -0,0 +1,20 @@
+    [ScriptImport]
+    [ScriptIgnoreNamespace]
+    [ScriptName("Object")]
+    public sealed class RestifyServerGetOptions {
```
Should have two ctors; a parameterless one and another accepting a params object[] nameValuePairs parameter.

theoutlander replied:
Changed

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServerOptions.cs:
```diff
@@ -0,0 +1,74 @@
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    [ScriptName("Object")]
+    public sealed class RestifyServerOptions {
```
Same ctors as the previous GetOptions.

theoutlander replied:
Changed

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServerOptions.cs:
```diff
@@ -0,0 +1,74 @@
+    [ScriptImport]
+    [ScriptName("Object")]
+    public sealed class RestifyServerOptions {
+        public string Test;
```
Is there an actual Test property on this class? If so, please move to proper location order-wise.

theoutlander replied:
Good catch....changed.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyServerOptions.cs:
```diff
@@ -0,0 +1,74 @@
+        ///         }
+        ///     });
+        /// </example>
+        public Func<int, int> ResponseTimeFormatter;
```
Sure this thing should return an int and not a string? A formatter sounds like something that should return a string.

theoutlander replied:
Interesting observation. I simply looked at the return value and picked int. In addition, I tried to play with an example on node and I wasn't ever able to get a formatted response time. So basically, I haven't been able to validate it.

I'm going to change it to string as it makes sense considering this is a formatter.

nikhilk commented on src/Libraries/Node/Node.Restify/RestifyThrottleOptions.cs:
```diff
@@ -0,0 +1,85 @@
+    [ScriptIgnoreNamespace]
+    [ScriptImport]
+    [ScriptName("Object")]
+    public sealed class RestifyThrottleOptions {
```
Ctors - like other options classes.

theoutlander replied:
Changed.

theoutlander commented:
Thanks again for your patience and the detailed CR. It is certainly helpful to have someone else look at this stuff especially when mapping to JS. Let me know if there are additional changes.

I've started to think about tutorials and created my first how-to page on enlisting in the CC branch and using it locally. Some reason my gh-pages branch is deleted, so I can't send a pull request for it. Once this checkin goes through, I'll delete and refork the s# branch so I can send that doc.

As I do more of these and get feedback, I'll get a good sense of how to structure tutorials as I move towards some more interesting ones. Next, I'm going to work towards a quick tutorial showcasing the AMD pattern and using SSLoader/RequireJS followed by a simple hello world type of example. It would be great if I can maintain continuity between tutorials.

theoutlander committed:
Additional fixes to initial Node-restify library port

theoutlander committed:
Removing files for initial node-restify commit

Removing files for initial node-restify commit

theoutlander commented:
I'm doing something wrong in squash and hence you're seeing several checkins :|. Will need to look at a good tutorial for it.

From what I understand I need to run **git rebase -i** and then pick the change sets and squash them into the latest one. However, I wonder if it got confused because I had already squashed an earlier one.

nikhilk commented:
I've seen various blogs/pages on using git rebase (see this one: https://help.github.com/articles/interactive-rebase) which suggest running:

```
git rebase -i master
```

It shouldn't matter that you've already squashed earlier changes once. Effectively that resulted in a new item in the change history at that point, that you'd be squashing with a subsequent rebase command.

If it works, then great. Otherwise, we'll just live with changes as they exist. If you try it again and it still doesn't help, just let me know and I'll merge as-is.

Oh, and thanks for the patience as well, while incorporating the comments! :)

theoutlander committed:
Merge branch 'cc' of https://github.com/theoutlander/scriptsharp into cc

theoutlander commented:
I seemed to have squashed it correctly, but ended up with a detached head. It asked me to sync with the remote and push changes. Now, some reason the commit shows as a merge. It does show 5 commits as part of it though, so maybe I got it right this time.

nikhilk commented:
Not sure what is happening - the intent was to see a single commit, while right now there are 7 commits in the pull request.
